### ISSUE TITLE
Riak CS does not correctly handle subresources when creating an AWS signature [JIRA: RCS-349]

### ISSUE THREAD

**Issue by bashoshaun** [basho/riak_cs#1292]
I have a customer that uses s3browser (s3browser.com) to interface with Riak CS.

Multi-part uploads fail for him.  This is due to the way Riak treats the uploadID subresource when creating an AWS signature.

Here's the steps taken and the results to provide the example:

User does a GET on a multi-part upload - this is the data captured:

```
GET /npo-training/020MB_txt?uploadId=YY0RyQnrTyiTOD3rCejqfQ%3D%3D HTTP/1.1
User-Agent: S3 Browser 5-7-3 http://s3browser.com
Authorization: AWS JQS79NNW-QPN_VZDYSCC:1e8MpKjK8F+ZRktLetv8IbWrmKs=
x-amz-date: Mon, 01 Feb 2016 14:43:31 GMT
Host: s3-emean-1.eecloud.nsn-net.net
```

Notice, that in the uploadID, part of it has been urlEncoded.  There's a %3D instead of an equals sign.

The signature is then created to authenticate with Riak CS.  Some clients, including s3cmd, would encode the following text:

```
GET



x-amz-date:Mon, 01 Feb 2016 14:43:31 GMT
/npo-training/020MB_txt?uploadId=YY0RyQnrTyiTOD3rCejqfQ==
```

As you can see, the uploadId was first URL decoded before signing.  In the user's example, it produced the hash of 1e8MpKjK8F+ZRktLetv8IbWrmKs= which matches the signature sent in the GET.

However, s3browser doesn't decode the URL.  It signs the following:

```
GET



x-amz-date:Mon, 01 Feb 2016 14:43:31 GMT
/npo-training/020MB_txt?uploadId=YY0RyQnrTyiTOD3rCejqfQ%3D%3D
```

This provides a different hash value, which means uploads are rejected.

Amazon has a description on how to properly create a signature [here](http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html).  Riak is treating the uploadId parameter as a _query string parameter_, which should be decoded from a URL encoding before being hashed.  However, Amazon considers an uploadID to be a _subresource_ which should be hashed in the encoding as it appears in the URL.

So when dealing with an uploadID, Riak CS should _not_ decode this part for the purposes of creating a signature (it can, of course, later decode it for other purposes).

Amazon define all the subresources as acl, lifecycle, location, logging, notification, partNumber, policy, requestPayment, torrent, uploadId, uploads, versionId, versioning, versions, and website.  I suspect all of these are URL decoded as well, although Riak CS probably doesn't use most of them.

**Comment by kuenishi** [comment]
This is a same issue as #1063, but thank you for reporting. This is the first client that does not work in v2 authentication.

**Comment by kuenishi** [comment]
Fixed by #1293 and will be in 2.1.2.

**Comment by Basho-JIRA** [comment]
Approved by Dave McCrory to create a 2.0 patch for Nokia.   

**_[posted via JIRA by Patricia Brewer]**_

**Comment by Basho-JIRA** [comment]
Previously fixed and included in Riak CS 2.1.2

Nokia patch is issue RCS-366 

**_[posted via JIRA by Ted Burghart]**_
